### PR TITLE
suggestion - feat: add feature flag loading in projects

### PR DESCRIPTION
**Pull request by srindom**
**What**
A simple draft for how we could load feature flags in Medusa projects.

- Feature flags are created as files in `src/loaders/feature-flags/*.ts`
- Each feature flag exports settings: key, description, default_value, etc. (subject to change)
- Feature flags can be controlled by 1) an env variable of the shape `MEDUSA_FF_{flag name}`, 2) a project flag set through `configModule.featureFlags.$`, 3) the default value specified by the settings as exported by the flag.

**Usage**
The loader registers a `featureFlagRouter` in the Awilix container; the registration contains a utility to check if a flag is enabled:

```
// from injected deps
constructor({ featureFlagRouter }) {
}

// in logic
myMethod() {
  if (featureFlagRouter.featureIsEnabled("flag_1")) {
    // do flag 1 stuff
  }
}
```

**Things to do**
- Make it possible to toggle migrations based on feature flags; TypeORM should ignore migrations behind a toggle
- Make it easy to create/delete flags; add commands to `medusa-dev`
  - `$ medusa-dev ff create my_feature_flag --description "used for my feature"`
  creates file at `/loaders/feature-flags/my-feature-flag.ts`
  - `$ medusa-dev ff delete my_feature_flag`
  - `medusa-dev` already knows where the root of the monorepo is on the dev machine so creating the file from a template will be simple.

@pKorsholm - I was bored this weekend so took a stab for fun; know you have done some research on this so feel free to pitch in if you have better approaches.

### CODE REVIEW

**Comment by pKorsholm** [comment]
Np, I will dump what I have below, but first a couple of thoughts. I definitely agree in the way you have done this, I also went for a projectconfig based approach.

One thing we should keep in mind is how do design this in a way where we can continually validate the old integration tests along with the integration tests for the new feature.  So how we can change either the projectConfig or write files into the `/feature-flags` directory. 

I went for a strategy based approach instead rather than a loader in order to support that if people want to use something else they should, I dont think it's strictly necessary though. 

I have my initial implementation here: https://github.com/medusajs/medusa/pull/1724 (ignore the mess in integration testing)

**Comment by pKorsholm** [comment]
**Initial write-up**
For an initial implementation there are generally two ways to go down. Either using a file-based feature flag system or using a database-based.

Since we wish to use *Release toggles* and don’t have a need to dynamically switch flags on and off we can utilize a static file approach in the local medusa instance. The requirement to restart the server when changing feature flags wouldn’t present an issue. When a feature is under development it can be added to the file of choice with a flag indicating wether or not this feature is enabled. This file could be a markup file or a json file that can be parsed by loaders when starting the medusa instance or the feature flag information could be part of the project config in the medusa-config file.

A different approach to feature flags could be to add an additional database table. Adding a new database table would represent a bigger investment in how feature flags are introduced into the system. It would also allow feature flags to be changed dynamically which as discussed is out of scope. 

In my opinion the best solution would be one that is simple and works for our use-case, while at the same time offers flexibility for the users of to implement their own custom solutions, wether a third party service or just a custom solution. An obvious candidate for this would be to include a `featureFlagStrategy` with the sole responsibility of, when given a featureflag, returning true if it’s enabled or false otherwise. This would provide the aforementioned flexibility to implement a 3rd-party feature flag provider.

Due to how feature flags should ideally be used, a positive response when checking if a flag is enabled should yield the new feature to be executed. In turn this means that a negative response should imply that the default behavior is used for features under development. With that in mind, we should consider how to handle non-existing feature-flags. I think we should pass all non-positives as false, including non-existing flags. In this way, if we chose a file based method of feature configuration we don’t need to extend the medusa-starter with an additional markup file. Instead since no configuration is passed, by default all feature flags will be disabled.

**My suggestion**

I think we should create this as a strategy as well, in order to facilitate that people wanting to use special feature flag software such as: [https://launchdarkly.com/](https://launchdarkly.com/). For that we should create an interface, and I think we only need something like this: 

```tsx
interface IFeatureFlagService {
	isSet(flagName: string): boolean
}
```

Initially I think we should go with a file-based approach to feature flags since it’s the easiest and we can always change our approach at a later point. Also it’s less obtrusive to create a db table than to remove it again.

I think we initially could include the featureflags in the `projectConfig` in `medusa-config.js` . In that way when we parse the config we can include the featureflags and dependency inject them into the corresponding strategy. The format could be along these lines: 

```tsx
module.exports = {
  projectConfig: {
    ...
  },
  plugins,
  featureFlags: {
    "medusa:batchJob": true
  }
};
```

The important part is not how the flags are formatted as it can be decided later, but that this is just a `Record<string, boolean>` which can be passed on to the `FeatureFlagService` . 

The service implementation would be very simple, like so:

```tsx
class FeatureFlagStrategy {
  private current_: Map<string, boolean>

  constructor({ configModule }) {
    this.current_ = new Map(
      Object.entries(configModule.featureFlags).map(([k, v]) => [k, !!v])
    )
  }

  isSet(flagname: string): boolean {
    return !!(this.current_.has(flagname) && this.current_.get(flagname))
  }
}

export default FeatureFlagStrategy
```

When using the `featureFlagStrategy` it would either be dependency injected or resolved like all other strategies we use, Example: using feature flags for API endpoints:

```tsx

export default (app, container, config) => {
	app.use("/admin", route)

  const adminCors = config.admin_cors || ""
  
	...

	const featureFlagService =	req.scope.resolve("featureFlagService")
	
	appRoutes(route)
  productRoutes(route)

	if(featureFlagService.isSet("medusa:batchJob"))
	{
		batchRoutes(route)
	}

	userRoutes(route)
  regionRoutes(route)

	...
}
```

**Comment by pKorsholm** [comment]
## Using Feature flags

Featureflags are controlled and implemented as described by Seb above (key based or environment variable based). This is comment is purely meant to discuss the use of the featureflags. 

Featureflags as they are implemented can be used to control behavior in two different ways, as general purpose if statements, and on the route level (using middleware @adrien2p). Furthermore, when a feature is under development migrations and changes to the datamodel should also be marked in order to not provoke unnecessary migrations.

**Route level** 

To block single routes or entire paths in the api, the middleware `featureFlagEnabled` can be used as so: 

```tsx
app.use("/batch-jobs", featureFlagEnabled("test_flag"), route)
```

This will ensure that if the featureflag with the flag key “test_flag” is not enabled all routes in it will return a general 404 error. If the flag is enabled the routes are accessible.

The middleware can also be applied to specific routes such as `create-batch-job` .

 

**General use**

For general use the featureflag utility is registered in awilix and can be resolved as so: 

```tsx
const featureFlagRouter = req.scope.resolve("featureFlagRouter") as FlagRouter

if (featureFlagRouter.featureIsEnabled("test_flag")){
	// guarded behavior
} else {
	// default behavior
}
```

In other cases the featureflag router can be obtained by constructor injection as all other services.

### Data model features

In order to develop the datamodel that a feature uses without migrating it unintentionally the datamodel should be disabled on the migration level as well as the appropriate model level. This prevents breaking the server if either is missing, as well as unintentionally applying a migration that is under development.

**Disabling migrations**

All the loading of migrations has been updated in the pr to support the filtering of migrations that are part of features which are not yet ready to be deployed. This filtering is based on an export that should be added to the newly created migration file when the feature is under development. The field to export should be called `featureFlag` as shown below. 

When running `medusa migrations run` or the integration tests this migration will be ignored unless specified by the environment, `medusa-config.js` or it being enabled by default. 

```tsx
import { MigrationInterface, QueryRunner } from "typeorm"

export const featureFlag = "test_flag"

export class addBatchJobModel1649775522087 implements MigrationInterface {
  name = "addBatchJobModel1649775522087"

  public async up(queryRunner: QueryRunner): Promise<void> {
		...
```

When disabling migrations it’s important to also address the corresponding model and/or model properties due to typeorm failing when there is a discrepancy between the loaded entities and the migrations. 

**Disabling model classes (entire database tables)** 

Typeorm uses metadata about each class in order to model the relationship from class to database table. This metadata is established by the class decorator `@Entity()` that we has annotated all classes in the `models` directory with. 

This doesn’t change for features under development. However it does require a different annotation in order to only load the entities when the feature flag is enabled. 

When adding models to a feature under development, the following decorator should be used along with the relevant featureflag as so: 

```tsx
import {
  FeatureFlagEntity,
	...
} from "../utils/db-aware-column"

@FeatureFlagEntity("test_flag")
export class BatchJob extends SoftDeletableEntity {
```

This decorator ensures that the entity is not loaded unless the corresponding featureflag is set.

**Disabling database model properties**

If the migration in question relates to adding columns to the database there has been added two column decorators as well. One for regular columns (what would otherwise be annotated with `@Column` and one for multiple columns in case additional descriptors are used. 

For annotating regular columns the `@FeatureFlagColumn` can be used. It acts as a regular column but doesn’t provide any metadata when the entity is loaded by Typeorm. This ensures that the field is not registered as a db column.

```tsx
@FeatureFlagColumn("test_flag", { default: true })
```

For decorating a field with multiple columns `@FeatureFlagColumn` can be used. Much like the above `@FeatureFlagColumn` it takes a featureflag for which the definition is a valid column. The second argument is a list of decorators to apply if the featureflag is active. 

```tsx
@FeatureFlagColumn("test_flag", [Column({ default: true })])
```

**Comment by pKorsholm** [comment]
## Testing

### Unit testing

**Services**

In services the featureflag router can be dependency injected along with other entities in the service. This means it’s easy to create a `new` featureflag router and pass it to the constructor to unit test the desired behavior.

**API Layer**

Testing the API layer is a bit different. To test endpoints with certain flags in the api layer, use the `flags` options for the request to include the enabled feature-flag.

```jsx
await request(
	"GET",
	`/admin/...`,
	{
		adminSession: {
			...
		},
		flags: ["batch_jobs"],
	}
)
```

### Integration testing

When testing it’s pivotal that we can test both the features guarded by feature-flags as well as the default features with the behavior of disabled feature flags. 

In order to setup an integration testing suite I have created the utility `startServerWithEnvironment` . This utility sets whatever environment variables is passed to it and runs `initDb` and `startServer` and returns the medusa-process as well as the connection.

In order to validate features that are guarded by featureflags the  `startServerWi...` method sets the environment based on the object passed along to the method. This allows us to run integration tests with the featureflags set. 

```tsx
const env = { MEDUSA_FF_CREATE_BATCHES: "true" }
const [process, connection] = await startServerWithEnvironment(cwd, env)
```

File: integration-tests/helpers/start-server-with-environment.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    process.env[key] = value
+  })
+
+  const dbConnection = await initDb({
+    cwd,
+  })
+
+  const medusaProcess = await setupServer({
+    cwd,
+    verbose: true,
```
Comment (srindom) [inline integration-tests/helpers/start-server-with-environment.js:15]: todo: we would likely want to control this on a test-by-test basis.

suggestion: change startServerWithEnv to take an object:
```
startServerWithEnvironment({ cwd, env, verbose }) {
```

File: packages/medusa/src/loaders/feature-flags/test-flag.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export default {
```
Comment (srindom) [inline packages/medusa/src/loaders/feature-flags/test-flag.ts:1]: todo: delete this file

File: packages/medusa/src/utils/db-aware-column.ts [hunk 3]
```diff
@@ -72,3 +82,67 @@ export function DbAwareColumn(columnOptions: ColumnOptions): PropertyDecorator {
...
+  const featureFlagRouter = featureFlagsLoader(configModule)
+
+  if (!featureFlagRouter.featureIsEnabled(featureFlag)) {
+    return (): void => {
+      // noop
+    }
+  }
+
+  return Column(columnOptions)
+}
```
Comment (srindom) [inline packages/medusa/src/utils/db-aware-column.ts:104]: todo: move this to its own file instead of db-aware-column

Comment (srindom) [reply packages/medusa/src/utils/db-aware-column.ts:104]: Same goes for all the other feature flag related work here

**Review by srindom — COMMENTED** [review]
Love it! Have a few changes for you :)

File: integration-tests/helpers/start-server-with-environment.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+const setupServer = require("./setup-server")
+const { initDb } = require("./use-db")
+
+const startServerWithEnvironment = async ({ cwd, verbose, env }) => {
+  if (env) {
+    Object.entries(env).forEach(([key, value]) => {
+      process.env[key] = value
+    })
+  }
```
Comment (srindom) [inline integration-tests/helpers/start-server-with-environment.js:9]: thought: one super tricky thing with this approach is that if people forget to reset `process.env = OLD_ENV` in `afterAll` or something similar; you might end up in some tricky debugging situations. 

Do you think that we could maybe do:
```suggestion
const OLD_ENV = process.env
  if (env) {
    Object.entries(env).forEach(([key, value]) => {
      process.env[key] = value
    })
  }
```

and then after `await initDb` do `process.env = OLD_ENV`? 

This way we are effectively only setting the env vars during the setup of the server and we don't require devs to remember to clean up after their tests.

File: packages/medusa/src/loaders/feature-flags/index.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        flagConfig[flagSettings.key] = isTruthy(
+          projectConfigFlags[flagSettings.key]
+        )
+        break
+      default:
+        if (logger) {
+          logger.info(
+            `Using flag ${flagSettings.key} with default value ${flagSettings.default_val}`
+          )
+        }
```
Comment (srindom) [inline packages/medusa/src/loaders/feature-flags/index.ts:69]: todo: we shouldn't log this as this should be the case that normal users experience. If we start telling them about feature flags they will likely be confused.

File: packages/medusa/src/utils/db-aware-column.ts (hunk 3 above)
Comment (srindom) [reply packages/medusa/src/utils/db-aware-column.ts:102]: This should be removed from here as it is now in a new file :)

**Review by srindom — COMMENTED** [review]
@pKorsholm - as discussed maybe move this to your own branch and I will close this PR :)

**Comment by srindom** [comment]
Closed in favor of #1724

**Pull request closed without merging**